lib/Serialization/SerializeDoc.cpp

```diff
@@ -712,55 +703,43 @@ writer.write<uint32_t>(data.X.Column);
     return USRWriter.getNewUSRId(OS.str());
   }
 
-  Optional<DeclLocationsTableData> getLocData(Decl *D) {
-    auto *File = D->getDeclContext()->getModuleScopeContext();
-    auto Locs = cast<FileUnit>(File)->getBasicLocsForDecl(D);
-    if (!Locs.hasValue())
-      return None;
-    DeclLocationsTableData Result;
-    llvm::SmallString<128> AbsolutePath = Locs->SourceFilePath;
-    llvm::sys::fs::make_absolute(AbsolutePath);
-    Result.SourceFileOffset = FWriter.getTextOffset(AbsolutePath.str());
-    Result.DocRangesOffset = DocWriter.getDocRangesOffset(D,
-      llvm::makeArrayRef(Locs->DocRanges));
-#define COPY_LINE_COLUMN(X)                                                   \
-Result.X.Line = Locs->X.Line;                                                 \
-Result.X.Column = Locs->X.Column;
-    COPY_LINE_COLUMN(Loc)
-    COPY_LINE_COLUMN(StartLoc)
-    COPY_LINE_COLUMN(EndLoc)
-#undef COPY_LINE_COLUMN
-    return Result;
-  }
-
   bool shouldSerializeSourceLoc(Decl *D) {
     if (D->isImplicit())
       return false;
     return true;
   }
 
   bool walkToDeclPre(Decl *D) override {
-    SWIFT_DEFER {
-      assert(USRWriter.peekNextId() * sizeof(DeclLocationsTableData)
-             == Buffer.size() &&
-            "USR Id has a one-to-one mapping with DeclLocationsTableData");
-    };
     // .swiftdoc doesn't include comments for double underscored symbols, but
     // for .swiftsourceinfo, having the source location for these symbols isn't
     // a concern because these symbols are in .swiftinterface anyway.
     if (!shouldIncludeDecl(D, /*ExcludeDoubleUnderscore*/false))
       return false;
     if (!shouldSerializeSourceLoc(D))
       return true;
-    // If we cannot get loc data for D, don't proceed.
-    auto LocData = getLocData(D);
-    if (!LocData.hasValue())
+
+    auto *File = D->getDeclContext()->getModuleScopeContext();
+    auto Positions = cast<FileUnit>(File)->getBasicPositionsForDecl(D);
```

Comment:
This is essentially just inlining getLocData to avoid the copy into the (now removed) DeclLocationsTableData. It had this cast previously and I assumed it was correct. If it wasn't though, it would imply we're serializing a Decl that has no FileUnit. Do you know when that could happen?

Note that this brings up another part of the serialization that I left as is. From what I can tell, the .swiftsourceinfo serialization was built similar to .swiftdocinfo. The docinfo serializes `DeclCommentTableData` directly into the on-disk hash table values, hence the need for `DeclCommentTableData`. I wasn't sure if there was a good reason that the sourceinfo doesn't do that, so I left as is but removed the unnecessary copy (mostly to avoid defining more slightly-different-to-basicsourceinfo structs) .

Reply:
> Do you know when that could happen?

No, I just see code that gets a module scope context and assumes it's a file.  So I'm asking how do we know it's a FileUnit?  If the answer is that we would never serialize something outside a file unit, great.

> I wasn't sure if there was a good reason that the sourceinfo doesn't do that

I don't know either. Maybe @nkcsgexi  would know.